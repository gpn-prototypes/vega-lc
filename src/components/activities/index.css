.ActivitiesWidget {
  position: absolute;

  margin: 16px;
  padding: 8px;

  width: 40%;
  max-width: 316px;
  height: 840px;

  background-color: var(--color-bg-default);

  &__Title {
    margin: 0;

    font-size: 16px;
    font-style: normal;
    line-height: 150%;
  }

  & .VegaTree {
    padding: 0;
  }
}
