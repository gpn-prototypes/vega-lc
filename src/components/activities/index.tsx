import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IconSearch, TargetData, Text, TextField, Tree } from '@gpn-prototypes/vega-ui';

import {
  setActivitiesDraggingElements,
  setActivitiesRef,
  setSearchString,
} from '../../redux-store/activities/actions';
import {
  getActivitiesNodeList,
  getIsAutoFocus,
  getSearchStringValue,
} from '../../redux-store/activities/selectors';

import { cnActivities } from './cn-activities';
import { BlueLineSvg } from './icons';

import './index.css';

const icons = {
  'blue-line': BlueLineSvg,
};

export const ActivitiesWidget: React.FC = (): React.ReactElement => {
  const dispatch = useDispatch();

  const activitiesRef = useRef(null);
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    dispatch(setActivitiesRef(activitiesRef));
  }, [dispatch, activitiesRef]);

  const searchString = useSelector(getSearchStringValue);
  const activities = useSelector(getActivitiesNodeList(searchString));
  const autoFocus = useSelector(getIsAutoFocus);

  const handleSearch = (args: { value: string | null }): void => {
    dispatch(setSearchString(args.value));
  };

  const handleDragStart = (transferringElems: TargetData[]): void => {
    dispatch(setActivitiesDraggingElements(transferringElems));
  };

  const handleDragEnd = (): void => {
    dispatch(setActivitiesDraggingElements([]));
  };

  return (
    <div ref={activitiesRef} className={cnActivities()}>
      <div className={cnActivities('Head')}>
        <Text className={cnActivities('Title').toString()}>Мероприятия</Text>

        <TextField
          autoFocus={autoFocus}
          className={cnActivities('Search').toString()}
          leftSide={IconSearch}
          size="s"
          inputRef={inputRef}
          type="input"
          onChange={handleSearch}
          value={searchString}
          placeholder="Поиск"
        />
      </div>

      <Tree
        icons={icons}
        nodeList={activities || []}
        withVisibilitySwitcher={false}
        withDropZoneIndicator={false}
        showIndentGuides={false}
        onDragStart={handleDragStart}
        onDragEnd={handleDragEnd}
      />
    </div>
  );
};
