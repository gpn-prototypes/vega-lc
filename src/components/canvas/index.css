.CanvasWidget {
  position: absolute;
  top: 0;
  right: 0;

  width: 100vw;
  height: calc(100vh - 30px);
}
