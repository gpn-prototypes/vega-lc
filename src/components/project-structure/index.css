.ProjectStructure {
  position: relative;

  height: 100%;

  & > .VegaTree {
    overflow-y: auto;

    height: 100%;
  }

  &__NoStructure {
    display: flex;
    justify-content: center;

    height: 100%;

    & > p {
      margin: 0;
      padding: var(--space-xs) 0;

      font-size: var(--size-text-s);

      text-align: center;
    }
  }
}
