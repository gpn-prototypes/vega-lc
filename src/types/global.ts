import ResizeObserver from 'resize-observer-polyfill';

declare global {
  interface Window {
    ResizeObserver: typeof ResizeObserver;
  }
}
