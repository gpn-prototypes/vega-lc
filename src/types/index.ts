export interface Identity {
  getToken(): Promise<string>;
}
