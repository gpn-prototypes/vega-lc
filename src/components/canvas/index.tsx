import React, { useContext, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Canvas, CanvasUpdate, Change, useTimeout } from '@gpn-prototypes/vega-ui';
import { CanvasView as CanvasViewEntity } from '@gpn-prototypes/vega-ui/dist/components/canvas/entities/CanvasView';

import './index.css';

import { cnCanvasWidget } from '@/components/canvas/cn-canvas';
import { ProjectContext } from '@/react-context/providers';
import {
  mapDropEventToRelatedAction,
  setCanvasElements,
  setCanvasViewRef,
  syncCanvasState,
  toggleStepEditor,
} from '@/redux-store/logic-constructor/actions';
import { getCanvasElements } from '@/redux-store/logic-constructor/selectors';
import { canvasActionsForImmediateSync } from '@/utils/constants/canvas-actions-to-sync';

export const CanvasWidget: React.FC = () => {
  const [changes, setChanges] = useState<CanvasUpdate[]>([]);

  const { projectId } = useContext(ProjectContext);

  const canvasViewRef = useRef<CanvasViewEntity | null>(null);

  const canvasElements = useSelector(getCanvasElements);

  const dispatch = useDispatch();

  const canvasViewRefSetter = (view: CanvasViewEntity): void => {
    if (canvasViewRef.current) return;

    canvasViewRef.current = view;

    dispatch(setCanvasViewRef(canvasViewRef));
  };

  useTimeout(2000, () => {
    if (changes.length) {
      dispatch(syncCanvasState(changes[changes.length - 1]));

      setChanges([]);
    }
  });

  const updateTree = (change: Change): void => {
    const { state, update } = change;
    const { type } = update;

    dispatch(setCanvasElements(state));

    if (update.type === 'select') {
      if (update.selected?.type === 'event') {
        dispatch(toggleStepEditor(true));

        return;
      }

      if (update.selected?.type === 'item') {
        dispatch(toggleStepEditor(false));

        return;
      }
    }

    if (type === 'unselect') {
      dispatch(toggleStepEditor(false));

      return;
    }

    if (canvasActionsForImmediateSync[type]) {
      dispatch(syncCanvasState(update));

      return;
    }

    if (update.type === 'drop-event') {
      const { intersectionId, position } = update;

      dispatch(mapDropEventToRelatedAction({ intersectionId, position }));

      return;
    }

    setChanges([...changes, update]);
  };

  return (
    <div className={cnCanvasWidget()}>
      <Canvas
        canvasViewAccessor={canvasViewRefSetter}
        state={canvasElements}
        onChange={updateTree}
        savedScreenId={projectId}
      />
    </div>
  );
};
