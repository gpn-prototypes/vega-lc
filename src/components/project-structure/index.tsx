import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { TargetData, Tree } from '@gpn-prototypes/vega-ui';

import { BlueLineSvg, OrangeLineSvg, RedLineSvg } from './icons';

import { setProjectStructureDraggingElements } from '@/redux-store/project-structure/actions';
import { getProjectStructureNodeList } from '@/redux-store/project-structure/selectors';

const icons = {
  'blue-line': BlueLineSvg,
  'orange-line': OrangeLineSvg,
  'red-line': RedLineSvg,
};

export const ProjectStructureWidget: React.FC = (): React.ReactElement => {
  const projectStructure = useSelector(getProjectStructureNodeList);
  const dispatch = useDispatch();

  const handleDragStart = (transferringElems: TargetData[]): void => {
    dispatch(setProjectStructureDraggingElements(transferringElems));
  };

  const handleDragEnd = (): void => {
    dispatch(setProjectStructureDraggingElements([]));
  };

  return (
    <Tree
      onDragStart={handleDragStart}
      onDragEnd={handleDragEnd}
      icons={icons}
      isContextMenuEnable
      nodeList={projectStructure || []}
    />
  );
};
