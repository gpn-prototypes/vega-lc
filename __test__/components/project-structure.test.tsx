import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { TreeItem } from '@gpn-prototypes/vega-tree';
import ResizeObserver from 'resize-observer-polyfill';

import '../../src/types/global';

import { ProjectStructureWidget } from '../../src/components/project-structure';
import getStore from '../../src/redux-store';

describe('Geological Exploration', () => {
  let mockProjectStructure: TreeItem[] = [];

  beforeEach(() => {
    mockProjectStructure = [
      {
        name: 'mock',
        id: '1',
        nodeList: [],
      },
    ];
  });

  test('рендерится без ошибок', () => {
    window.ResizeObserver = ResizeObserver;

    const dom = renderer
      .create(
        <Provider store={getStore()}>
          <ProjectStructureWidget projectStructure={mockProjectStructure} />
        </Provider>,
      )
      .toJSON();
    expect(dom).toMatchSnapshot();
  });
});
