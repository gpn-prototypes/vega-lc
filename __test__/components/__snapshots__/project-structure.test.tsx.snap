// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geological Exploration рендерится без ошибок 1`] = `
<div
  className="ProjectStructure"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <p>
      Структура проекта отсутствует
    </p>
  </div>
</div>
`;
